File: 2487-optimal-partition-of-string/2487-optimal-partition-of-string.py
class Solution:
    def partitionString(self, s: str) -> int:

        last_seen = [-1] * 26
        count = 0
        # substring helps to count occurence of same characters in different substring
        substring = 0

        for i in range(len(s)):
            if last_seen[ord(s[i]) - ord('a')] >= substring:
                count += 1
                substring = i
            last_seen[ord(s[i]) - ord('a')] = i
        return count+1

File: 2487-optimal-partition-of-string/2487-optimal-partition-of-string.java
import java.io.*;
import java.util.Arrays;

class Solution {
    public int partitionString(String s) {

        int count = 1;
        int substring = 1;
        int[] lastSeen = new int[26];
        char c;

        for(int i =1; i<=s.length();i++)
        {
            c = s.charAt(i-1);
            // This ensures that the new substring is accounted
            if(lastSeen[((int) c)-((int) 'a')] >= substring)
            {
                count += 1;
                substring = i;
            }
            lastSeen[((int) c)-((int) 'a')] = i;
        }

        return count;
    }
}